### PR TITLE
upgrade rules for NVMeofGW monitors and gateways

### PR DESCRIPTION
**Pull request by leonidc**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/mon/NVMeofGwMap.h [hunk 1]
```diff
@@ -102,29 +107,31 @@ class NVMeofGwMap
...
-    using ceph::decode;
-    DECODE_START(1, bl);
-    decode(epoch, bl);
-
-    decode(created_gws, bl);
-    decode(fsm_timers, bl);
-    DECODE_FINISH(bl);
-  }
+    void encode(ceph::buffer::list &bl, uint64_t features) const {
+      uint8_t version = 1;
```
Comment (athanatos) [inline src/mon/NVMeofGwMap.h:116]: version isn't used anywhere else in this function, just pass a 1 into ENCODE_START.

File: src/mon/NVMeofGwMap.h [hunk 2]
```diff
@@ -102,29 +107,31 @@ class NVMeofGwMap
...
+      ENCODE_START(version, 1, bl);
```
Comment (athanatos) [inline src/mon/NVMeofGwMap.h:117]: The indentation on this section is wrong.  Please re-indent this and every other line in this commit to match my cleanup.

File: src/mon/NVMeofGwMap.h [hunk 3]
```diff
@@ -102,29 +107,31 @@ class NVMeofGwMap
     const NvmeGroupKey& group_key);
 
 public:
+
```
Comment (athanatos) [inline src/mon/NVMeofGwMap.h:110]: This is a stray newline.  Please check every line in this commit for similar accidental additions.

File: src/mon/NVMeofGwMon.cc [hunk 4]
```diff
@@ -429,76 +446,76 @@ bool NVMeofGwMon::preprocess_beacon(MonOpRequestRef op)
...
+bool NVMeofGwMon::prepare_beacon(MonOpRequestRef op) {
+    auto m = op->get_req<MNVMeofGwBeacon>();
 
   dout(20) << "availability " <<  m->get_availability()
-	   << " GW : " << m->get_gw_id()
-	   << " osdmap_epoch " << m->get_last_osd_epoch()
-	   << " subsystems " << m->get_subsystems() << dendl;
+        << " GW : " << m->get_gw_id()
+        << " osdmap_epoch " << m->get_last_osd_epoch()
+        << " subsystems " << m->get_subsystems() << dendl;
```
Comment (athanatos) [inline src/mon/NVMeofGwMon.cc:455]: This chunk changes appears to be an accidental revert of my cleanup PR (changes the indent but doesn't change the content).  Please check over this entire PR for other examples and remove them.

File: src/mon/NVMeofGwMon.cc [hunk 5]
```diff
@@ -429,76 +446,76 @@ bool NVMeofGwMon::preprocess_beacon(MonOpRequestRef op)
...
-	propose = true;
-	goto set_propose;
+          gw_availability_t::GW_AVAILABLE) {
+        dout(4) << " Warning :GW marked as Available in the NVmeofGwMon "
+            "database, performed full startup - Force gw to exit!"
+            << gw_id <<dendl;
+        avail = gw_availability_t::GW_UNAVAILABLE;
+        // Monitor performs Force Failover for this GW in process_gw_map_gw_down
+      }
+      else if (pending_map.created_gws[group_key][gw_id].performed_full_startup
```
Comment (athanatos) [inline src/mon/NVMeofGwMon.cc:492]: This chunk appears to be re-introducing formatting I fixed.  Please do not do that.

File: src/mon/NVMeofGwTypes.h [hunk 6]
```diff
@@ -157,35 +157,58 @@ struct NqnState {
     : nqn(_nqn), ana_state(_ana_state)  {}
   NqnState(
     const std::string& _nqn, const SmState& sm_state,
-    const NvmeGwMonState & gw_created)
-    : nqn(_nqn)  {
+    const NvmeGwMonState & gw_created,  bool encode_gmap_last)
```
Comment (athanatos) [inline src/mon/NVMeofGwTypes.h:160]: No.  If you need to modify the structures on encode, make a copy and do it in encode.

**Review by athanatos — CHANGES_REQUESTED** [review]
I'll take another look once you've fixed the stray style reverts.

File: src/messages/MNVMeofGwMap.h [hunk 7]
```diff
@@ -20,8 +20,8 @@
 
 class MNVMeofGwMap final : public Message {
 private:
-  static constexpr int VERSION = 1;
-
+  static int constexpr VERSION = 2;
```
Comment (ronen-fr) [inline src/messages/MNVMeofGwMap.h:23]: That's a pretty unusual order of declarations. I surely  would not have *fixed* 'constexpr int' into 'int constexpr'

File: src/mon/NVMeofGwSerialize.h [hunk 8]
```diff
@@ -17,6 +17,7 @@
 #undef dout_prefix
 #define MODULE_PREFFIX "nvmeofgw "
 #define dout_prefix *_dout << MODULE_PREFFIX << __PRETTY_FUNCTION__ << " "
+#define MAX_SUPPORTED_ANA_GROUPS 16 //used for support backward compatibility
```
Comment (ronen-fr) [inline src/mon/NVMeofGwSerialize.h:20]: maybe rephrase 'for support'

File: src/mon/NVMeofGwTypes.h [hunk 9]
```diff
@@ -157,35 +157,58 @@ struct NqnState {
...
+	    ? gw_exported_states_per_group_t::GW_EXPORTED_OPTIMIZED_STATE
+	        : gw_exported_states_per_group_t::GW_EXPORTED_INACCESSIBLE_STATE;
+        state_pair.second =
+            gw_created.blocklist_data.at(state_itr.first).osd_epoch;
+        ana_state.push_back(state_pair);
+        i++;
+      }
+    }
+    else {
+#define MAX_SUPPORTED_ANA_GROUPS 16 // ths define is used for backward compatibility
```
Comment (ronen-fr) [inline src/mon/NVMeofGwTypes.h:193]: strange location for a '#define'. After an 'else' - but takes effect anyway.

File: src/test/test_nvmeof_mon_encoding.cc [hunk 10]
```diff
@@ -50,9 +50,9 @@ void test_NVMeofGwMap() {
   pending_map.created_gws[group_key]["GW2"].nonce_map[2] = new_nonces;
   dout(0) << " == Dump map before Encode : == " <<dendl;
   dout(0) << pending_map << dendl;
-
+  uint64_t features = 4540701547738038271;
```
Comment (ronen-fr) [inline src/test/test_nvmeof_mon_encoding.cc:53]: maybe using an hex number to initialize? would save the reader the need to fire up a calculator...
